src/csharp/Grpc.Core/Internal/AsyncCall.cs

```diff
@@ -510,6 +510,7 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
 
             lock (myLock)
             {
+                readingDone = true;
```

Comment:
Ok, so here's the description of the exact C core behavior:

Behavior at end of stream:
1. if status HAS NOT been requested (recv_status_on_client, recv_close_on_server have NOT been posted), all messages will be available up to the end of the stream, which will be designated with a NULL received message
2. if status HAS been requested, AND the status is OK, all messages will be available up to the end of the stream
3. if status HAS been requested AND the status is NOT OK, some messages may be discarded by core

As you can see, cutting off the read stream is only fine if the received status is NOT OK (case 3). In other cases, we can't really implement an early disposal of the call as we would risk throwing away messages sent by the server.